hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemListStatusIterator.java

```diff
@@ -0,0 +1,190 @@
+    Path rootPath = new Path(rootPathStr);
+    getFileSystem().create(rootPath);
+    RemoteIterator<FileStatus> fsItr = fs.listStatusIterator(rootPath);
+    Assertions.assertThat(fsItr.hasNext())
+        .describedAs("hasNext returns true for file").isTrue();
+  }
+
+  @Test
+  public void testHasNextForIOException() throws Exception {
+    final AzureBlobFileSystem fs = getFileSystem();
+    String rootPathStr = "testRoot5";
+    Path rootPath = new Path(rootPathStr);
+    getFileSystem().mkdirs(rootPath);
+    ListStatusRemoteIterator fsItr = (ListStatusRemoteIterator) fs
+        .listStatusIterator(rootPath);
+    Thread.sleep(1000);
+
+    String exceptionMessage = "test exception";
+    setPrivateField(fsItr, ListStatusRemoteIterator.class, "ioException",
+        new IOException(exceptionMessage));
+    setPrivateFinalField(fsItr, ListStatusRemoteIterator.class,
+        "iteratorsQueue", new ArrayBlockingQueue<Iterator>(1));
+
+    Assertions.assertThatThrownBy(() -> fsItr.hasNext()).describedAs(
+        "When ioException is not null and queue is empty exception should be "
+            + "thrown").isInstanceOf(IOException.class)
+        .hasMessage(exceptionMessage);
+  }
+
+  private void setPrivateField(Object obj, Class classObj, String fieldName,
+      Object value) throws NoSuchFieldException, IllegalAccessException {
+    Field field = classObj.getDeclaredField(fieldName);
+    field.setAccessible(true);
+    field.set(obj, value);
+  }
+
+  private void setPrivateFinalField(Object obj, Class classObj,
+      String fieldName, Object value)
+      throws NoSuchFieldException, IllegalAccessException {
+    Field field = classObj.getDeclaredField(fieldName);
+    field.setAccessible(true);
+    Field modifiersField = Field.class.getDeclaredField("modifiers");
+    modifiersField.setAccessible(true);
+    modifiersField.setInt(field, field.getModifiers() & ~Modifier.FINAL);
+    field.set(obj, value);
+  }
+
+  private List<String> createFiles(int numFiles, String rootPathStr,
+      String filenamePrefix)
+      throws ExecutionException, InterruptedException, IOException {
+    final List<Future<Void>> tasks = new ArrayList<>();
+    final List<String> fileNames = new ArrayList<>();
+    ExecutorService es = Executors.newFixedThreadPool(10);
+    final Path rootPath = new Path(rootPathStr);
+    for (int i = 0; i < numFiles; i++) {
+      final Path filePath = new Path(rootPath, filenamePrefix + i);
+      Callable<Void> callable = new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          getFileSystem().create(filePath);
+          fileNames.add(makeQualified(filePath).toString());
+          return null;
+        }
+      };
+      tasks.add(es.submit(callable));
+    }
+    for (Future<Void> task : tasks) {
+      task.get();
+    }
+    es.shutdownNow();
+    return fileNames;
+  }
+
+  private AzureBlobFileSystemStore getAbfsStore(FileSystem fs)
```

Comment:
this is why you need a list callback. This is low-level field abuse. Better to have a {{getAbfsStoreForTesting()}} call, at the very least

Reply:
Done